workflow/controller/operator.go

```diff
@@ -1785,6 +1785,12 @@ func (woc *wfOperationCtx) executeTemplate(ctx context.Context, nodeName string,
 		}
 	}
 
+	if node.Fulfilled() {
+		if processedTmpl.Synchronization != nil {
```

Comment:
I think you could merge the two if expressions into one statement

Reply:
what about adding these two checks to `.Release` - i.e. have it refuse to release if the node is unfulfilled or the synchronization is nil? the less code we add to woc, the better